### ISSUE TITLE
CLI flag to launch Pulumi program under a debugger

### ISSUE THREAD

**Issue by swgillespie** [pulumi/pulumi#1372]
Since we get to use ✨ real code ✨ in Pulumi programs, sometimes you just need to look at your code in a debugger. It would be cool if we could do this from the pulumi CLI.

Language hosts could optionally respond to an RPC request to launch a given program under a debugger where the debugger is listening to a given port.

**Comment by ellismg** [comment]
Moving this out.  We'll probably do this around the same time that we try to support doing more Pulumi stuff from within the language host itself (e.g. when we support managing the lifecycle of a stack in the source program itself, so you can write tests using your favorite framework).

**Comment by mikhailshilkov** [comment]
With .NET support, this should get up in the queue IMO, and .NET folks **love** their F5-debugging.

**Comment by lukehoban** [comment]
A quick update on this.

In Python, .NET and Go it is actually already possible to debug Pulumi programs today - though requires a little bit of manual work.

As one example of this - here's steps for Python and VS Code:
1. Follow these instructions https://code.visualstudio.com/docs/python/debugging#_attach-to-a-local-script.  
2. Run `pulumi up` to launch your deployment.  
3. It will appear to hang.  
4. Look in your process list for the `python` process that is running your program.  
5. Get its PID, and use `Python: Attach` in VS Code to attach to that PID.

Similar steps should work for .NET and Go and using other debuggers for those runtimes.

Today - this is not yet possible with Node.js as the equivalent of step (1) above for Node.js is to pass `--debug-brk` flag to the Node.js process - and `pulumi` does not pass that by default and does not yet allow users to opt in to passing it.  We will be enabling that soon.

We will also be looking into automating a bit more of this process where possible so that the end to end experience is smoother here.

**Comment by joeduffy** [comment]
@lukehoban Is this in 2.0?

**Comment by johnnyhuy** [comment]
@lukehoban yep, can confirm it works with .NET debugging via attaching to the *application* process.

Just had to either wait for a debugger or create a debugger process.

```c#
while (!Debugger.IsAttached)
{
    Thread.Sleep(100);
}

// OR

Debugger.Launch();
```

[Reference](https://stackoverflow.com/questions/361077/how-to-wait-until-remote-net-debugger-attached)

**Comment by jonathh21** [comment]
> @lukehoban yep, can confirm it works with .NET debugging via attaching to the _application_ process.
> 
> Just had to either wait for a debugger or create a debugger process.
> 
> ```cs
> while (!Debugger.IsAttached)
> {
>     Thread.Sleep(100);
> }
> 
> // OR
> 
> Debugger.Launch();
> ```
> 
> [Reference](https://stackoverflow.com/questions/361077/how-to-wait-until-remote-net-debugger-attached)

Whilst this works, and it does, because i just tested it, everything is waiting to complete so you can't actually see anything :)

**Comment by cmclaughlin** [comment]
Is there a way to debug with Python outside of VS Code?

**Comment by cmclaughlin** [comment]
https://pypi.org/project/rpdb/ works, but I don't see a way to get tab-completion working... maybe it's best to try PyCharm or VS Code...

**Comment by ctwise** [comment]
Typescript use-case: security tightened down egress whitelist and pulumi script stopped working. It didn't timeout, it just hung. It took some time to figure it out. Time that would have been saved if we could just attach a debugger to it.

**Comment by lukehoban** [comment]
For those using Go - there are a couple additional steps needed.

First, you will need a `time.Sleep` in your code to give you a chance to attach before the code you want to debug is run.  Unfortunately Go does not have a "wait for attach" API, which means this is a little clunkier than in other languages.

Second, since `go run` strips debugging symbols, you will need to manually `go build` and then put a section like this in `Pulumi.yaml`:

```yaml
runtime:
    name: go
    options:
        binary: goapp
```

Adjust the `binary: ` configuration to point at the binary you have built locally, and then `pulumi up` will deploy using that binary and you can attach to it with your debugger.

**Comment by mikhailshilkov** [comment]
> you will need a `time.Sleep` in your code to give you a chance to attach

FWIW, I used to add a loop with a short sleep inside and a condition like `!stop`, then change the value of `stop` manually when the debugger is attached. This way you don't have to be faster than the single large delay.

**Comment by Malet** [comment]
@lukehoban it's possible to pass it in using an environment variable:
```shell
NODE_OPTIONS='--inspect-brk' pulumi up
```

**Comment by dnlfrst** [comment]
> For those using Go - there are a couple additional steps needed.
> 
> First, you will need a `time.Sleep` in your code to give you a chance to attach before the code you want to debug is run. Unfortunately Go does not have a "wait for attach" API, which means this is a little clunkier than in other languages.
> 
> Second, since `go run` strips debugging symbols, you will need to manually `go build` and then put a section like this in `Pulumi.yaml`:
> 
> ```yaml
> runtime:
>     name: go
>     options:
>         binary: goapp
> ```
> 
> Adjust the `binary: ` configuration to point at the binary you have built locally, and then `pulumi up` will deploy using that binary and you can attach to it with your debugger.

I am using [delve](https://github.com/go-delve/delve) as `go` debugger. While I was able to `pulumi up` with a manually build (i.e., `go build`) binary, I could not manage to attach the debugger to it. `time.Sleep` was reasonably set to one minute and it seemed that `pulumi up` did halt at the breakpoint but the debugger did not report so. 

@lukehoban what debugger do you use and is there anything else I need to pay attention to apart from what you already mentioned?

**Comment by gdodd1977** [comment]
Is there any update on a clean way to do this? The Attach to process isn't working for me with .net app.

**Comment by gitfool** [comment]
@gdodd1977 it's a little rough around the edges right now due to requiring the `pulumi` executable to drive things. I use the following `StackBase` class:
```c#
public abstract class StackBase : Stack
{
    protected StackBase()
    {
        bool.TryParse(Environment.GetEnvironmentVariable("PULUMI_DEBUG"), out var debug);
        bool.TryParse(Environment.GetEnvironmentVariable("PULUMI_DRY_RUN"), out var dryRun);

        if (debug && dryRun)
        {
            // Log.Debug("Waiting for debugger...");
            while (!Debugger.IsAttached)
            {
                Thread.Sleep(100);
            }
            // Log.Debug("Waited for debugger");
        }
    }
}
```
To attach during preview I set the `PULUMI_DEBUG` env var to `true` - which is not an official Pulumi env var, that's just how I named it - before running `pulumi up` from the command-line, and I can attach to the _child process_ of `dotnet.exe` from Visual Studio 2019:

![image](https://user-images.githubusercontent.com/750121/93930259-3032f880-fd71-11ea-8a0b-f5c91423d83a.png)

**Comment by joeduffy** [comment]
For Node.js, I found the easiest way to launch with a debugger is

1) Add a [`debugger;` statement](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/debugger) statement where you want to ultimately break (to skip through a bunch of internal machinery)

2) Launch as @Malet suggests above:

    `NODE_OPTIONS='--inspect-brk' pulumi up`

3) Attach the debugger -- for instance, [here is how to do it with VS Code](https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_attaching-to-nodejs)

4) Hit "Go" until you land on the `debugger;` statement inserted during step 1

**Comment by rienafairefr** [comment]
Adding my 2cts, not sure why that works but it seems to, by using the `node-terminal` type in launch.json
```
{

    "version": "0.2.0",
    "configurations": [
         {
            "name": "Deploy run",
           "command": "pulumi up", 
           "type": "node-terminal"
         }
    ]
}
```
typescript program, pulumi 3.10.3, node 12, vscode 1.59.1
I can put a breakpoint in the index.ts, hit the debug, et voilà.

**Comment by aureq** [comment]
For people using Python and VSCode, here is an alternative.

1. Install `debugpy` as part of your environment.
2. Add `import debugpy` to your Pulumi app
3. In your VSCode project add a new launch configuration and select `Python` and then `Remote Attach`. Keep the default (`localhost` and `5678`)
4. Add the code below to your code

```python
debugpy.listen(("localhost", 5678))
max_wait_s = 30
while max_wait_s >= 0:
    if debugpy.is_client_connected():
        break

    time.sleep(1)
    max_wait_s -= 1
```

To start debugging, simply run `pulumi up` or `pulumi preview` from the command line and after a few seconds, hit `F5` in VSCode.

VSCode should stop at the first encountered breakpoint.

**Comment by mritd** [comment]
## For people using GoLand and Go (This idea may apply to other languages):

- 1. First of all, Pulumi directly executes the binary file/script when it starts, and Pulumi sets specific environment variables when executing the binary file (I guess other languages may be the same).
- 2. Secondly, Pulumi allows you to set the location of the executable file.
- 3. I think any modern language should support "remote debugging"; for GoLand, it uses dlv for remote debugging;

After understanding the execution method, we only need to create a simple script, then execute the [dlv](https://github.com/go-delve/delve) debugging command in the script, and finally turn on the remote Debug function of GoLand, we can happily Coding.

## Below is the detailed tutorial:

### This is my project structure:

```sh
~/o/p/xxxx ❯❯❯ tree
.
├── Pulumi.yaml
├── common.go
├── debug.sh
├── deployment.go
├── go.mod
├── go.sum
├── goapp
├── main.go
└── service.go

0 directories, 9 files
```

### Compile Code to bin file

For GoLand's remote Debug, we need to compile our code using the following command:

```sh
# this will generate an executable file with debug symbols (goapp)
go build -o goapp -gcflags "all=-N -l" .
```

### Create debug script

We create a bash script to execute the dlv command:

⚠️ Note: Please replace `./goapp` with a real absolute path (I'm not sure if you don't do this will cause problems, but using an absolute path must be fine)

```sh
#!/bin/bash
dlv --listen=:2345 --headless=true --api-version=2 --accept-multiclient exec ./goapp
```

Save the above code as `debug.sh`, and execute `chmod +x debug.sh` to increase its executable permissions.

### Update `Pulumi.yaml`

We need to update `Pulumi.yaml `to let Pulumi execute our `debug.sh` script:

⚠️ The following `binary` configuration should also use the absolute path of `debug.sh`

```yaml
name: test
description: Debug example
runtime:
  name: go
  options:
    binary: /Users/bleem/ops/pulumi/xxxx/debug.sh
```

### Create GoLand Remote Debug configuration

![image](https://user-images.githubusercontent.com/13043245/138240664-50b1f5b5-808c-43b6-a9e1-872ff50c19b1.png)


### Start debugging

Execute the `pulumi up` command in the terminal, which will block pulumi and display a line of log:

```sh
Type                 Name       Plan     Info
pulumi:pulumi:Stack  xxxxx-dev           API server listening at: [::]:2345
```

Finally, click the Debug button on GoLand to let us "find bugs, fix bugs, create bugs".

![image](https://user-images.githubusercontent.com/13043245/138241743-c5af50cb-9323-46c7-84c9-a0ecad0fa470.png)

**Comment by zivyatziv** [comment]
> For people using Python and VSCode, here is an alternative.
> 
> 1. Install `debugpy` as part of your environment.
> 2. In your VSCode project add a new launch configuration and select `Python` and then `Remote Attach`. Keep the default (`localhost` and `5678`)
> 3. Add the code below to your code
> 
> ```python
> import debugpy
> debugpy.listen(("localhost", 5678))
> max_wait_s = 30
> while max_wait_s >= 0:
>     if debugpy.is_client_connected():
>         break
> 
>     time.sleep(1)
>     max_wait_s -= 1
> ```
> 
> To start debugging, simply run `pulumi up` or `pulumi preview` from the command line and after a few seconds, hit `F5` in VSCode. VSCode should pause the execution at the first encountered breakpoint.
> 
> If no debugger is attached after 30 seconds, execution will continue.

I would like to suggest a better approach which pauses the code and waits for a client to attach:
```python
import debugpy

debugpy.listen(("0.0.0.0", 5678))
print("debugpy is listening, attach by pressing F5 or ►")

debugpy.wait_for_client()
print("Attached to debugpy!")
```

Since I am working with `venv` in pulumi, I would often get this error:
```logos
Unable to open '__main__.py': Unable to read file '/pulumi/__main__.py' (Error: Unable to resolve non-existing file '/pulumi/__main__.py').
```

You can solve this issue by changing the remote path in the `launch.json` like so:
```json
"pathMappings": [
    {
        "localRoot": "${workspaceFolder}/",
        "remoteRoot": "${workspaceFolder}/"     # <-- default is empty which results in the root project directory 
    }
]
```

**Comment by harryttd** [comment]
I'm using Typescript and this works for me:

```shell
❯ code --version
1.63.0
7db1a2b88f7557e0a43fec75b6ba7e50b3e9f77e
x64
```

`package.json`:

[![package.json screenshot with vscode debugger prompt][1]][1]

Notice the "debug" prompt above the scripts section. In between lines 5 and 6. You can click that and vscode will open up a debug console. Works perfectly.
https://stackoverflow.com/a/70661424/6379084

  [1]: https://i.stack.imgur.com/LuSSr.png

**Comment by secustor** [comment]
I can confirm simply starting it in a debugger as @harryttd  described, works in Webstorm too.

**Comment by ghostsquad** [comment]
> For those using Go - there are a couple additional steps needed.
> 
> First, you will need a `time.Sleep` in your code to give you a chance to attach before the code you want to debug is run. Unfortunately Go does not have a "wait for attach" API, which means this is a little clunkier than in other languages.
> 
> Second, since `go run` strips debugging symbols, you will need to manually `go build` and then put a section like this in `Pulumi.yaml`:
> 
> ```yaml
> runtime:
>     name: go
>     options:
>         binary: goapp
> ```
> 
> Adjust the `binary: ` configuration to point at the binary you have built locally, and then `pulumi up` will deploy using that binary and you can attach to it with your debugger.

Is it possible to use any of the `--config*` flags to dynamically set/merge these debugging properties? It's not clear how these flags should be used.

![image](https://user-images.githubusercontent.com/903488/159545209-b9f81a1f-19de-4ca8-8787-8a9ec77939cf.png)

**Comment by Frassle** [comment]
Note that nodejs programs can be quite easily debugged using vscodes auto-attach system (https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_auto-attach)

Turn on auto attach and then run `pulumi preview/up` in the vscode terminal and vscode will see the new node process and attach to it.

**Comment by zivyatziv** [comment]
> Note that nodejs programs can be quite easily debugged using vscodes auto-attach system (https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_auto-attach)
> 
> Turn on auto attach and then run `pulumi preview/up` in the vscode terminal and vscode will see the new node process and attach to it.

Or you can create a new terminal with the type `JavaScript debug terminal` And that's it :) 
![image](https://user-images.githubusercontent.com/39673349/171111418-a4edcc49-58fe-4c48-976b-1d0da50abca1.png)

**Comment by cristichiru** [comment]
I have found a simpler way to debug pulumi go project in VSCode.

There is no need to build the project upfront nor add `options.binary`.

1. As @lukehoban suggested, add to the beginning of your pulumi.Run func: `time.Sleep(20 * time.Second)`
2. Set a breakpoint somewhere in your code
3. Run `pulumi up` or `pulumi preview`. Wait a couple of seconds to allow `go build` (invoked by pulumi) to finish
4. In VSCode, assuming you have an "Attach to Process" configuration, start debugging (<kbd>F5</kbd>)

   A list of processes should be displayed, filter to `pulumi-go`
   ![image](https://user-images.githubusercontent.com/28685998/207848607-2bb80a9b-b52e-44b7-ba26-022faccb37bd.png)

![image](https://user-images.githubusercontent.com/28685998/207849294-531ddac6-398f-4321-ac0f-97a1f97c303e.png)

**Comment by Pithikos** [comment]
Anyone got this working with IPython?

E.g.

```
import pulumi
from pulumi_gcp import storage

# Create a GCP resource (Storage Bucket)
bucket = storage.Bucket('my-bucket', location="eu")

import IPython; IPython.embed(using=False)
```

The debug point is then totally skipped when running Pulumi
```
Diagnostics:
  pulumi:pulumi:Stack (trade-engine-staging):
    Python 3.9.16 (main, Dec  7 2022, 01:11:51)
    Type 'copyright', 'credits' or 'license' for more information
    IPython 8.10.0 -- An enhanced Interactive Python. Type '?' for help.
    In [1]: Do you really want to exit ([y]/n)?

Resources:
    + 1 created
    1 unchanged
```

**Comment by weijiany** [comment]
There is a way for `intelliJ`.

1. Using this command to start inspector `NODE_OPTIONS='--inspect-brk' pulumi preview -s [your stack name]`. The default server address is `localhost:9229`, you could find it in this [document](https://nodejs.org/en/docs/guides/debugging-getting-started).
2. Opening your project via intelliJ. Following the config:
    <img width="1805" alt="image" src="https://user-images.githubusercontent.com/26241124/233391605-5084ab44-9f4b-4df0-b5cc-0dc8893bc442.png">
3. Adding your breakpoint and then `⌘ + D` start your debug mode. 🎉

**Comment by sanmai-NL** [comment]
@swgillespie This issue/discussion would better be refined into issues/topics for specific languages.

**Comment by ami-descope** [comment]
> There is a way for `intelliJ`.
> 
> 1. Using this command to start inspector `NODE_OPTIONS='--inspect-brk' pulumi preview -s [your stack name]`. The default server address is `localhost:9229`, you could find it in this [document](https://nodejs.org/en/docs/guides/debugging-getting-started).
> 2. Opening your project via intelliJ. Following the config:
>    <img alt="image" width="1805" src="https://user-images.githubusercontent.com/26241124/233391605-5084ab44-9f4b-4df0-b5cc-0dc8893bc442.png">
> 3. Adding your breakpoint and then `⌘ + D` start your debug mode. 🎉

thank you so much!!!! this is life changing!

**Comment by tazik561** [comment]
> 1. NODE_OPTIONS='--inspect-brk' pulumi preview -s

How did you use  ? In the first step when I tap to enter the pulumi starts to work!  since there is problem in my pulumi codes after starting I will get error. I want to debug it but this guild does not work for me.
I am using dot net

**Comment by Frassle** [comment]
> I am using dot net

NODE_OPTIONS is a nodejs specific option. The simplest thing I've seen for dotnet is to add a call to `System.Diagnostics.Debugger.Launch()` at the start of your program.

**Comment by tazik561** [comment]
> > I am using dot net
> 
> NODE_OPTIONS is a nodejs specific option. The simplest thing I've seen for dotnet is to add a call to `System.Diagnostics.Debugger.Launch()` at the start of your program.

Thanks for response , If I run it without pulumi I got this error : 
```
Unhandled exception. System.InvalidOperationException: Program run without the Pulumi engine available; re-run using the `pulumi` CLI
   at Pulumi.Deployment..ctor(RunnerOptions runnerOptions)
   at Pulumi.Deployment.<>c__119`1.<RunAsync>b__119_0()
   at Pulumi.Deployment.CreateRunnerAndRunAsync(Func`1 deploymentFactory, Func`2 runAsync)
   at RR.ClientPortalApi.Infra.Program.<Main>()
```

How can I use with pulumi preview ? I mean If I run pulumi preview by terminal , How can I debug it ?

**Comment by Frassle** [comment]
Add a call to [`System.Diagnostics.Debugger.Launch()`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.debugger.launch?view=net-7.0) at the start of your program and then run `pulumi preview`.

pulumi will run the program with the environment needed for the sdk to work, and the `Launch()` method should trigger a debugger attach dialog when it's hit.

**Comment by tazik561** [comment]
> 

I am using mac, I was using both but not working :
```
    public class SharedStack : Stack
    {
#if DEBUG
        private const string BuildConfiguration = "Debug";
#else
        private const string BuildConfiguration = "Release";
#endif

        public SharedStack() => CreateResources();

    

        private void CreateResources()
        {
            System.Diagnostics.Debugger.Break();
```

**Comment by EronWright** [comment]
Summarizing how to attach across the SDKs, such that the program waits for the client to attach.
1. nodejs - use `--inspect-brk` flag when spawning `node`, attach by pid.
2. golang - use `dlv exec` command to run the program, attach by port.
3. python - use `debugpy` in entrypoint script to listen and wait, attach by port. Must install `debugpy` into venv.
4. csharp - poll `System.Diagnostics.Debugger.IsAttached`  in entrypoint script to wait, attach by pid.
5. java - use `jdwp` JVM args, attach by port.

**Comment by yann-soubeyrand** [comment]
Hello, for Go the new approach seems to use DAP (`dlv dap`).

**Comment by lukehoban** [comment]
We've added official docs for Debugging Pulumi Programs, summarizing a lot of the notes and suggestions users have left in comments above.  

https://www.pulumi.com/docs/using-pulumi/debugging/

There are guides specifically for:
* [Python Debugging](https://www.pulumi.com/docs/using-pulumi/debugging/#debugging-python-programs)
* [.NET Debugging](https://www.pulumi.com/docs/using-pulumi/debugging/#debugging-net-programs)
* [Go Debugging](https://www.pulumi.com/docs/using-pulumi/debugging/#debugging-go-programs)
* [Java Debugging](https://www.pulumi.com/docs/using-pulumi/debugging/#debugging-java-programs)

Users should be able to debug and step through their Pulumi programs in any of these languages already today using those guides.

We will leave this issue open to track some enhancements we would still like to make to automate more of this workflow, as well as provide even more integration with some common IDE debuggers.

**Comment by lukehoban** [comment]
We have now shipped debugging support in the `pulumi` CLI, as well as a new [Pulumi VS Code extension](https://www.pulumi.com/blog/pulumi-vscode-extension/) with support for easy debugging on Pulumi programs in all languages.

The documentation at https://www.pulumi.com/docs/using-pulumi/debugging/ is now updated to leverage both of those new capabilities, making debugging with Pulumi even easier, whether in VSCode or in other tools.

Given these two latest releases, we're now ready to close this issue out! 🎉
